### PR TITLE
ingester: Add `/ingester/tsdb_metrics` endpoint for returning tenant-specific TSDB metrics.

### PR DESCRIPTION
**Pull request by pstibrany**
#### What this PR does
This PR adds `/ingester/tsdb_metrics` endpoint to ingester for returning tenant-specific TSDB metrics. This is not intended for generic metrics scraping, but can be used as a troubleshooting tool.

#### Checklist

- [na] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/util/metrics_helper_test.go [hunk 1]
```diff
@@ -770,7 +771,7 @@ func TestUserRegistries_RemoveUserRegistry_HardRemoval(t *testing.T) {
 	mainRegistry := prometheus.NewPedanticRegistry()
 	mainRegistry.MustRegister(tm)
 
-	// Soft-remove single registry.
+	// Hard-remove single registry.
```
Comment (replay) [inline pkg/util/metrics_helper_test.go:774]: This looks unrelated to the rest of the PR? Is it just a random comment fix, or an accident?

**Review by replay — APPROVED** [review]
Looks great, thanks!

File: pkg/util/metrics_helper_test.go (hunk 1 above)
Comment (pstibrany) [reply pkg/util/metrics_helper_test.go:774]: This is fix for copy&pasted comment. `true` parameter on next line indicates "hard" removal.

**Comment by pstibrany** [comment]
I will wait with merging this until there are more reviews (or for a few days) to see if people are for or against the idea of such endpoint.

File: pkg/api/api.go [hunk 2]
```diff
@@ -271,6 +272,7 @@ func (a *API) RegisterIngester(i Ingester, pushConfig distributor.Config) {
 	a.RegisterRoute("/ingester/flush", http.HandlerFunc(i.FlushHandler), false, true, "GET", "POST")
 	a.RegisterRoute("/ingester/shutdown", http.HandlerFunc(i.ShutdownHandler), false, true, "GET", "POST")
 	a.RegisterRoute("/ingester/push", push.Handler(pushConfig.MaxRecvMsgSize, a.sourceIPs, a.cfg.SkipLabelNameValidationHeader, i.PushWithCleanup), true, false, "POST") // For testing and debugging.
+	a.RegisterRoute("/ingester/tsdb_metrics", http.HandlerFunc(i.UserRegistryHandler), false, false, "GET")
```
Comment (pracucci) [inline pkg/api/api.go]: Don't we want auth to be enabled here, given it's a per-tenant endpoint?

```suggestion
	a.RegisterRoute("/ingester/tsdb_metrics", http.HandlerFunc(i.UserRegistryHandler), true, false, "GET")
```

File: pkg/ingester/ingester.go [hunk 3]
```diff
@@ -2482,3 +2483,22 @@ func allOutOfBounds(samples []mimirpb.Sample, minValidTime int64) bool {
...
+	}
+
+	reg := i.tsdbMetrics.regs.GetRegistryForUser(userID)
+	if reg == nil {
+		http.Error(w, "user registry not found", http.StatusNotFound)
+		return
+	}
+
+	promhttp.HandlerFor(reg, promhttp.HandlerOpts{
+		ErrorHandling: promhttp.HTTPErrorOnError,
```
Comment (pracucci) [inline pkg/ingester/ingester.go:2501]: Have you considered setting `DisableCompression: true` and then enable compression when we `RegisterRoute()`? Would be consistent with other API endpoints registered by Mimir.

File: pkg/util/metrics_helper.go [hunk 4]
```diff
@@ -675,6 +675,25 @@ func (r *UserRegistries) Registries() []UserRegistry {
...
+	r.regsMu.Lock()
+	defer r.regsMu.Unlock()
+
+	for idx := 0; idx < len(r.regs); idx++ {
+		if user != r.regs[idx].user {
+			continue
+		}
+
+		if r.regs[idx].reg == nil {
+			continue
```
Comment (pracucci) [inline pkg/util/metrics_helper.go:689]: [nit] Can you add a comment to explain why? I had to look at the rest of the code to understand it. It's because the registry to set to nil on soft deletion, but the user may have a new registry registered.

**Review by pracucci — APPROVED** [review]
LGTM! I just left a couple of nits.

File: pkg/api/api.go (hunk 2 above)
Comment (pstibrany) [reply pkg/api/api.go]: Fixed in https://github.com/grafana/mimir/pull/3923/commits/4d7f523a4954af8ada0f051b680dd47098af3afe

File: pkg/ingester/ingester.go (hunk 3 above)
Comment (pstibrany) [reply pkg/ingester/ingester.go]: I haven't, but sounds like a good idea. Fixed in 8a7e1947bf7f033876031d6264f3ff72647483c4.

File: pkg/util/metrics_helper.go (hunk 4 above)
Comment (pstibrany) [reply pkg/util/metrics_helper.go:689]: Yes, I was thinking about that use-case that you describe, but I just realized that we also set `user` to `""` on soft-deletion, so this if condition can never be true (unless one passes `""` parameter to the function). I will remove it.

Comment (pstibrany) [reply pkg/util/metrics_helper.go]: https://github.com/grafana/mimir/pull/3923/commits/863f9afc24f96a1a77d3b2fbee8ea0da57dedb2b

**Comment by pstibrany** [comment]
Two approvals, no objections, merging :) Thanks for reviews!
